### PR TITLE
feat: LLD allow users to use Ledger Vault as a signer

### PR DESCRIPTION
**Pull request by drakoFukayu**
### 📝 Description

Reopening the PR 3287 https://github.com/LedgerHQ/ledger-live/pull/3287

### ❓ Context

- **Impacted projects**: `ledger-live-desktop` `ledger-live-common` `ledgerjs/hw-transport-http`
- **Linked resource(s)**: [VG-14376 ](https://ledgerhq.atlassian.net/browse/VG-14376)

### ✅ Checklist

- [ ] **Test coverage**
- [x] **Atomic delivery**
- [x] **No breaking changes**

### 📸 Demo

Demo

Between the record of this GIF, I updated some wording and added some Alerts:

Demo

### 🚀 Expectations to reach

_Please make sure you follow these [**Important Steps**](https://github.com/LedgerHQ/ledger-live/blob/develop/CONTRIBUTING.md#important-steps)._

_Pull Requests must pass the CI and be internally validated in order to be merged._

### CODE REVIEW

File: apps/ledger-live-desktop/src/renderer/components/VaultSignerTransport.tsx
Comment (gre) [inline apps/ledger-live-desktop/src/renderer/components/VaultSignerTransport.tsx:34]: `registerTransportModule` was not yet designed to be dynamically used. it probably works the first time, but the problem is that we don't have a way to "unregister" which will end up being registering multiple time each time the effect runs and could create race conditions.

I would suggest:
- to move `registerTransportModule` as a module definition level again, next to the others registerTransportModule (like the one we do in the live common setup logic). => it shouldn't be a problem as you will filter on the id
- as you need some extra information (host, token, workspace) in that transport, you may also need to have it part of the id. for instance `vault-transport?host=foo&token=bar&workspace=42`

File: libs/ledgerjs/packages/hw-transport-http/src/VaultTransport.ts
Comment (gre) [inline libs/ledgerjs/packages/hw-transport-http/src/VaultTransport.ts:9]: it would be better to make `VaultTransport` a dedicated package because it is not the scope of hw-transport-http i think.

an alternative idea is to add feature directly on the WebSocketTransport class, but it should not be breaking. it might be good to have an independent package that you can maintain independently (and that depends on hw-transport-http)

File: libs/ledger-live-common/src/hw/index.ts
Comment (gre) [inline libs/ledger-live-common/src/hw/index.ts:50]: i would prefer not to do this (see first comment) or if we want to go to such approach, the function should instead return a "unregister" function and it's user land responsability not to register twice.

**Review by gre — CHANGES_REQUESTED** [review]

**Comment by gre** [comment]
👋  we have recently merged https://github.com/LedgerHQ/ledger-live/pull/3455 in develop branch which reinforces the TypeScript checks, mostly related to modals and coin families. Could you rebase and check that everything still pass?

in the PR, i could see there is an impact on the way modals was defined, it is likely you will have to define a new modal in `modals/types.ts` and type the data it takes, then you should no longer expect a `name` generic prop from the modal code, but inline the name="YOUR_NAME".

Thanks, gre.

File: apps/ledger-live-desktop/src/renderer/components/VaultSignerTransport.tsx
Comment (flocks) [reply apps/ledger-live-desktop/src/renderer/components/VaultSignerTransport.tsx:34]: The problem is I need the transport to be registered each time `host/token/workspace` changes, that's why it's inside the `useEffect()`

**Comment by flocks** [comment]
Thx @gre I will have a look next week

File: apps/ledger-live-desktop/src/renderer/components/VaultSignerTransport.tsx
Comment (gre) [reply apps/ledger-live-desktop/src/renderer/components/VaultSignerTransport.tsx:34]: but then it could work if it's part of the id?

Comment (flocks) [reply apps/ledger-live-desktop/src/renderer/components/VaultSignerTransport.tsx:34]: I struggle to see what you mean tbh, if I register the vault transport [here](https://github.com/LedgerHQ/ledger-live/blob/develop/apps/ledger-live-desktop/src/renderer/live-common-setup.ts#L24) how can make it re-register each time `host/token/workspace` change?

File: libs/ledgerjs/packages/hw-transport-http/src/VaultTransport.ts
Comment (flocks) [reply libs/ledgerjs/packages/hw-transport-http/src/VaultTransport.ts:12]: I think you are right, vault transport should be a dedicated package!
I will work on that

File: apps/ledger-live-desktop/src/renderer/components/VaultSignerTransport.tsx
Comment (gre) [reply apps/ledger-live-desktop/src/renderer/components/VaultSignerTransport.tsx:32]: example https://github.com/LedgerHQ/ledger-live/blob/main/libs/ledger-live-common/src/load/speculos.ts#L471-L496

**Comment by gre** [comment]
👋  ESLint config has been unified on develop after we merged https://github.com/LedgerHQ/ledger-live/pull/3622 – this will improve the project consistency, remove eslint dependencies duplicates, and overall simplify the maintenance. This comes with a prettier library upgrade and unification of the prettierrc too, so you likely have conflicts with develop.


Hopefully, here is the procedure to easily fix this. **please proceed as follow**:

**Step 1: sync your branch with develop up to the config changes:**

2 cases

- if your branch were continuously rebased with develop and don't have any merge commit, you can do 
```
git rebase aa1e0ebc0a0498b4350754899b6a368fbd42e811
```
- otherwise prefer 
```
git merge aa1e0ebc0a0498b4350754899b6a368fbd42e811
```

**then, apply the config**

```
pnpm i
pnpm lint:fix --force
```

if there are remaining errors that lint couldn't fix, please fix them and `pnpm lint` again.

then commit in your branch.

```
git commit -a -m "chore: lint fixes"
```

**Step 2: finalize the rebase/merge with develop**

2 cases

- if your branch were continuously rebased with develop and don't have any merge commit, you can do
```
git pull --rebase origin develop
```
- otherwise prefer 
```
git merge origin/develop
```

you should now be up to date. 👍

---

*Thanks and sorry for the inconvenience,*
*gre*

File: apps/ledger-live-desktop/src/renderer/live-common-setup.ts
Comment (gre) [inline apps/ledger-live-desktop/src/renderer/live-common-setup.ts:39]: in future, the device mode should be provided by the transport module because it seems related to the transport used anyway. i think you can go with a hack that setDeviceMode it back in the usb transport module's open() but it's something we will need to track somewhere ( cc @juan-cortes @alexandremgo )

Comment (flocks) [reply apps/ledger-live-desktop/src/renderer/live-common-setup.ts:39]: Yep agree it feels weird to set it back in the `open` of an another transport, but I can think of a better solution  :/

**Review by gre — APPROVED** [review]

File: apps/ledger-live-desktop/src/renderer/live-common-setup.ts
Comment (flocks) [reply apps/ledger-live-desktop/src/renderer/live-common-setup.ts:44]: @gre just pushed this commit because i realized it was not working with a real device. It seems we actually never open a transport with `id=ipc` so the module was never picked up. May be it changed because I remember testing toggling on and off vault transport when I initially worked on that (weeks ago now).

**Comment by flocks** [comment]
@gre just in case you missed my [previous comment](https://github.com/LedgerHQ/ledger-live/pull/3523#discussion_r1226796494). And if you agree with that and re-approve, do you mind if I rebase and squash all commits on this branch into 1 single commit so it will ease potential future rebase onto develop, in case this PR is not merged right away?

**Comment by gre** [comment]
@flocks you can squash all commits.

**Review by gre — APPROVED** [review]
✔️ i did a smoke test on Mac and can still properly use a nano device.

**Pull request merged**
